ethcore/src/miner/service_transaction_checker.rs

```diff
@@ -0,0 +1,212 @@
+// Copyright 2017 Parity Technologies (UK) Ltd.
+// This file is part of Parity.
+
+// Parity is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Parity is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Parity. If not, see <http://www.gnu.org/licenses/>.
+
+use client::MiningBlockChainClient;
+use transaction::SignedTransaction;
+use util::{U256, Uint, Mutex};
+
+const SERVICE_TRANSACTION_CONTRACT_REGISTRY_NAME: &'static str = "service_transaction_checker";
+
+/// Service transactions checker.
+#[derive(Default)]
+pub struct ServiceTransactionChecker {
+	contract: Mutex<Option<provider::Contract>>,
+}
+
+impl ServiceTransactionChecker {
+	/// Try to create instance, reading contract address from given chain client.
+	pub fn update_from_chain_client(&self, client: &MiningBlockChainClient) {
+		let mut contract = self.contract.lock();
```

Comment:
Wouldn't it be easier to construct `provider::Contract` on each call without memoizing it?
You could probably just memoize the parsed ABI and then construct `provider::Contract` - in such case `explicit_do_call` would be unecessary.